server/services/aiOrder.js

```diff
@@ -0,0 +1,131 @@
+module.exports = class AIOrderService {
+    static CREATE_CARRIER_LOOP = "CREATE_CARRIER_LOOP";
+
+    constructor(carrierService, waypointService, starService, starUpgradeService) {
+        this.carrierService = carrierService;
+        this.waypointService = waypointService;
+        this.starService = starService;
+        this.starUpgradeService = starUpgradeService;
+        this.orderHandlers = {
+            CREATE_CARRIER_LOOP: this._handleCreateCarrierLoop.bind(this)
+        }
+    }
+
+    async processOrdersForPlayer(game, player, orders) {
+        if (!orders) {
+            return;
+        }
+
+        const newOrders = [];
+
+        for (let order of orders) {
+            let success = false;
+            try {
+                success = await this._performOrder(game, player, order.orderType, order.data);
+            } finally {
+                if (!success) {
+                    newOrders.push(order);
+                }
+            }
+        }
+
+        player.scheduledOrders = newOrders;
+    }
+
+    async _performOrder(game, player, orderType, data) {
+        const orderHandler = this.orderHandlers[orderType];
+        if (orderHandler) {
+            try {
+                return await orderHandler(game, player, data);
+            } catch (e) {
+                console.error(e);
+            }
+        }
+
+        return false;
+    }
+
+    async _handleCreateCarrierLoop(game, player, data) {
+        if (!data || !data.from || !data.to) {
+            return false;
+        }
+
+        const expenseConfig = game.constants.star.infrastructureExpenseMultipliers[game.settings.specialGalaxy.carrierCost];
+        const carrierCost = this.starUpgradeService.calculateCarrierCost(game, expenseConfig);
+
+        const carriersAtFrom = this.getAvailableCarriersAtStar(game, player, data.from);
+        const carriersAtTo = this.getAvailableCarriersAtStar(game, player, data.to);
+
+        let carrier;
+        let start;
+        let startAction;
+        let end;
+        let endAction;
+
+        if (carriersAtFrom && carriersAtFrom.length != 0) {
+            carrier = carriersAtFrom[0];
+            start = data.from;
+            end = data.to;
+            startAction = "collectAll";
+            endAction = "dropAll";
+        } else if (carriersAtTo && carriersAtTo.length != 0) {
+            carrier = carriersAtTo[0];
+            start = data.to;
+            end = data.from;
+            startAction = "dropAll";
+            endAction = "collectAll";
+        } else if (player.credits >= carrierCost) {
```

Comment:
What about carriers in transit?

We'll need to also check whether there's carrier travelling to or from the target stars otherwise the AI will create additional loops.